modules/java-modules/hdfs/src/main/java/org/syslog_ng/hdfs/HdfsDestination.java

```diff
@@ -231,28 +343,41 @@ public void onMessageQueueEmpty() {
          */
 
         logger.debug("Flushing hdfs");
+        lock.lock();
+        try {
         for (Map.Entry<String, HdfsFile> entry : openedFiles.entrySet()) {
-            FSDataOutputStream outputStream = entry.getValue().getFsDataOutputStream();
-            if (outputStream != null) {
-                try {
-                    outputStream.hflush();
-                    outputStream.hsync();
-                } catch (IOException e) {
-                    logger.debug(String.format("Flush failed on file %s, reason: %s", entry.getKey(), e.getMessage()));
-                }
+            HdfsFile hdfsfile = entry.getValue();
+            try {
+                hdfsfile.flush();
+            } catch (IOException e) {
+                logger.debug(String.format("Flush failed on file %s, reason: %s", entry.getKey(), e.getMessage()));
             }
         }
+        } finally {
+          lock.unlock();
+        }
     }
 
     @Override
     public void close() {
         logger.debug("Closing hdfs");
-        closeAll(archiveDir != null);
+
+        lock.lock();
+        try {
+        closeAll(true);
```

Comment:
hm... seems you want to always archive regardless if it is asked or not

Reply:
I mark this as resolved as it is duplicate of the above and either one would be changed or not.